staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,257 @@
+	config               *Config
+	backends             atomic.Value
+	webhookClientManager webhook.ClientManager
+}
+
+type syncedBackends map[types.UID]*Backend
+
+func (s syncedBackends) Names() []string {
+	names := []string{}
+	for _, backend := range s {
+		names = append(names, backend.configuration.Name)
+	}
+	return names
+}
+
+// ProcessEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEvents(events ...*auditinternal.Event) {
+	for _, b := range d.GetBackends() {
+		b.delegate.ProcessEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	// Shutdown channels are handled individually with the dynamic
+	// backend, message must be propagated out
+	go func() {
+		<-stopCh
+		d.disableUpdates()
+		d.stopAllBackends()
+	}()
+	var funcs []func() error
+	for _, b := range d.GetBackends() {
+		funcs = append(funcs, func() error {
+			return b.delegate.Run(b.stopChan)
+		})
+	}
+	return errors.AggregateGoroutines(funcs...)
+}
+
+// disableUpdates permanently disables the ability to add backends
+// this is used to shutdown gracefully
+func (d *dynamic) disableUpdates() {
+	d.backendUpdateMutex.Lock()
+}
+
+// stopAllBackends closes the stopChan for every backend to enable
+// goroutines to terminate gracefully then calls shutdown
+func (d *dynamic) stopAllBackends() {
+	for _, b := range d.GetBackends() {
+		if !isChanClosed(b.stopChan) {
+			close(b.stopChan)
+		}
+		b.delegate.Shutdown()
+	}
+}
+
+// Shutdown closes all backends stop channels and then runs their
+// subsequent shutdown methods
+func (d *dynamic) Shutdown() {
+	for _, b := range d.GetBackends() {
+		b.delegate.Shutdown()
+	}
+}
+
+// GetBackends retrieves current backends in a safe manner
+func (d *dynamic) GetBackends() syncedBackends {
+	return d.backends.Load().(syncedBackends)
+}
+
+// SetBackends retrieves the current backends in a safe manner
+func (d *dynamic) setBackends(backends syncedBackends) {
+	d.backends.Store(backends)
+}
+
+// addSink is called by the shared informer when a sink is added
+func (d *dynamic) addSink(sink *auditregv1alpha1.AuditSink) {
+	d.backendUpdateMutex.Lock()
+	defer d.backendUpdateMutex.Unlock()
+	b, err := d.createBackend(sink)
+	if err != nil {
+		glog.Errorf("could not add sink %s: %v", sink.Name, err)
+		return
+	}
+	backends := d.GetBackends()
+	backends[sink.UID] = b
+	d.setBackends(backends)
+	glog.V(2).Infof("added audit sink: %s", sink.Name)
+	glog.V(2).Infof("current audit sinks: %v", backends.Names())
+}
+
+// updateSink is called by the shared informer when a sink is updated
+func (d *dynamic) updateSink(oldSink, newSink *auditregv1alpha1.AuditSink) {
+	d.backendUpdateMutex.Lock()
+	defer d.backendUpdateMutex.Unlock()
+	b, err := d.createBackend(newSink)
+	if err != nil {
+		glog.Errorf("could not update sink %s: %v", oldSink.Name, err)
+		return
+	}
+	backends := d.GetBackends()
+	delete(backends, oldSink.UID)
```

Comment:
So one problem with this is it means the buffer will get reset during an update (potentially dropping a bunch of events). This isn't a problem if a common buffer is used instead (as long as shutdown waits for pending requests to complete).